neuralpredictors/training/tracking.py

```diff
@@ -146,3 +153,360 @@ def asdict(self, time_key="time", make_copy=True):
+        deep_update(self.log, new_log)
+
+    def start_epoch(self):
+        """ Start a new epoch. Initialize each accumulation with its default value. """
+        t = time.time()
+        self.time.append(t)
+        self.epoch += 1
+        self._initialize_epoch(self.log, self.objectives)
+
+    def log_objective(self, value: float, key: Tuple[str, ...] = ()):
+        """
+        Add a new entry to the logs
+        Args:
+            value: objective score
+            key: hierarchical key to match in `self.log`
+        """
+        if key:
+            self._log_objective_value(value, self.log, key)
+        else:
+            self._log_objective_callables(self.log, self.objectives)
+
+    def display_log(self, key=(), tqdm_iterator=None):
+        """
+        Display the current objective value of everything under `key`
+        Args:
+            key: Tuple describing the objective to display.
+                 This could be something like `("Training","img_classification","loss")`
+                 to display the current classification loss or something like `("Training","img_classification")`
+                 to display everything we save for image classifcation training.
+            tqdm_iterator: A tqdm object that the log could be displayed on.
+        """
+        # normalize (if applicable) and turn into np.arrays:
+        n_log = self._normalize_log(self.log)
+        # flatten a subset of the dictionary:
+        current_log = self._gather_log(n_log, key, index=-1)
+        if tqdm_iterator:
+            tqdm_iterator.set_postfix(**current_log)
+        else:
+            logger.info(current_log)
+
+    def check_isfinite(self, log: dict = None) -> bool:
+        """
+        Checks if all entries in `log` or `self.log` are finite.
+        Args:
+            log: dict that is recursively searched for infinite entries
+
+        Returns: True if all entries are finite.
+        """
+        if log is None:
+            log = self._normalize_log(self.log)
+        if isinstance(log, dict):
+            for k, l in log.items():
+                if not self._check_isfinite(l):
+                    return False
+        else:
+            return np.isfinite(log).any()
+        return True
+
+    def get_objective(self, log=None, key: Tuple[str, ...] = ()) -> np.array:
+        """
+        Args:
+            log: log to retrieve objective from
+            key: key to match with the log
+
+        Returns:
+            value array (across epochs and normalized) for a specific objective key
+        """
+        if log is None:
+            log = self._normalize_log(self.log)
+        if len(key) > 1:
+            return self.get_objective(log[key[0]], key[1:])
+        else:
+            return log[key[0]]
+
+    def get_current_objective(self, key) -> float:
+        return self.get_objective(self._normalize_log(self.log), key)[-1]
+
+    def get_current_main_objective(self, key) -> float:
```

Comment:
A bit confused as to why we need `key` here. Doesn't `self.main_objective` correspond to a single sequence of values in the log?

Reply:
See above (comment in the init).